test/api/comments/comment_test.rb

```diff
@@ -29,20 +29,80 @@ def test_student_post_comment
+      "comment" => "Hello World",
+      "has_attachment" => false,
+      "type" => "text",
+      "is_new" => false,
+      author: { "id" => user.id },
+      recipient: { "id" => tutor.id }
     }
 
     # check each is the same
-    assert_json_matches_model last_response_body, expected_response, ["comment", "has_attachment", "type", "is_new"]
+    assert_json_matches_model last_response_body, expected_response, %w(comment has_attachment type is_new)
     assert_json_matches_model last_response_body["author"], expected_response[:author], ["id"]
     assert_json_matches_model last_response_body["recipient"], expected_response[:recipient], ["id"]
   end
 
+  def test_replying_to_comments
+    project = Project.first
+    user = project.student
+    unit = project.unit
+    task_definition = unit.task_definitions.first
+    tutor = project.tutor_for(task_definition)
+
+    comment_data = { comment: "Hello World" }
+
+    # Post original comment and check that it was successful
+    post_json with_auth_token("/api/projects/#{project.id}/task_def_id/#{task_definition.id}/comments", user), comment_data
+    assert_equal 201, last_response.status
+
+    expected_response = {
+      "comment" => "Responding!",
+      "has_attachment" => false,
+      "type" => "text",
+      "is_new" => false,
+      "reply_to" => TaskComment.last.id,
+      author: { "id" => user.id },
+      recipient: { "id" => tutor.id }
+    }
+
+    # Student responding to self
+    comment_data = { comment: "Responding!", reply_to: TaskComment.last.id }
+    post_json with_auth_token("/api/projects/#{project.id}/task_def_id/#{task_definition.id}/comments", user), comment_data
+    assert_equal 201, last_response.status
+    assert_json_matches_model last_response_body, expected_response, %w(comment type is_new reply_to)
+
+    expected_response = {
+      "comment" => "Responding again!",
+      "has_attachment" => false,
+      "type" => "text",
+      "is_new" => false,
+      "reply_to" => TaskComment.last.id,
+      author: { "id" => user.id },
+      recipient: { "id" => tutor.id }
+    }
+    # Tutor responding to student
+    comment_data = { comment: "Responding again!", reply_to: TaskComment.last.id }
+    post_json with_auth_token("/api/projects/#{project.id}/task_def_id/#{task_definition.id}/comments", tutor), comment_data
+    assert_equal 201, last_response.status
+
+    # check each is the same
+    assert_json_matches_model last_response_body, expected_response, %w(comment type is_new reply_to)
+  end
+
+  def test_student_post_reply_to_invalid_comment
+    project = Project.first
+    user = project.student
+    unit = project.unit
+    task_definition = unit.task_definitions.first
+
+    comment_data = { comment: "Responding!", reply_to: -1 }
+
+    post_json with_auth_token("/api/projects/#{project.id}/task_def_id/#{task_definition.id}/comments", user), comment_data
+
+    assert_equal 404, last_response.status
+  end
+
```

Comment:
Can we also test:

- reply to comment not in unit - so valid comment id but should not be accessible
- reply to other students comment - so valid comment id but should not be accessible
- reply to other student comment in a group task (where valid [same team] and not valid [other team])
- also tutor reply to a different student comment in other student task - so valid comment, with access, but in unrelated student's task
- also what about reply to a comment in a different task for the same student - I assume this should not be allowed.

Reply:
These should all be enforced via active model validations.